File: Tree/104. Maximum Depth of Binary Tree /MaximumDepthOfBinaryTree.cpp
//
//  MaximumDepthOfBinaryTree.cpp
//  LeetCode-main
//
//  Created by jackma on 2022/3/2.
//  Copyright © 2022 JackMa. All rights reserved.
//

#include "MaximumDepthOfBinaryTree.hpp"

/**
 Given the root of a binary tree, return its maximum depth.

 Input: root = [3,9,20,null,null,15,7]
 Output: 3
 
 Input: root = [1,null,2]
 Output: 2
 */

/// use recursion to find max depth
static inline int _findTreeDepth(TreeNode *root, int depth) {
    if (root == NULL) {
        return depth;
    }
    int leftDepth = _findTreeDepth(root->left, depth+1);
    int rightDepth = _findTreeDepth(root->right, depth+1);
    return max(leftDepth, rightDepth);
}

int Solution::maxDepth(TreeNode *root) {
    return _findTreeDepth(root, 0);
}

int otherMaxDepth(TreeNode *root) {
    if (root == NULL) {
        return 0;
    }
    return 1+max(otherMaxDepth(root->left), otherMaxDepth(root->right));
}

void testMaxDepth() {
    vector<int> nums = {3,INT_MAX,7};
    cout << Solution().maxDepth(buildTree(nums)) << endl;
}

File: Tree/104. Maximum Depth of Binary Tree /MaximumDepthOfBinaryTree.swift
//
//  MaximumDepthOfBinaryTree.swift
//  LeetCode-Swift
//
//  Created by jackma on 2023/3/11.
//  Copyright © 2023 JackMa. All rights reserved.
//

import Foundation

/**
 https://leetcode.com/problems/maximum-depth-of-binary-tree
 
 Given the root of a binary tree, return its maximum depth.

 A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
 */

extension Solution {
    func MaximumDepthOfBinaryTree(_ root: TreeNode?) -> Int {
        guard let root = root else { return 0 }
        return root.findTreeDepth(1)
    }
}
